packages/flutter_tools/lib/src/ios/xcode_debug.dart

```diff
@@ -0,0 +1,498 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+import 'package:process/process.dart';
+
+import '../base/common.dart';
+import '../base/file_system.dart';
+import '../base/io.dart';
+import '../base/logger.dart';
+import '../base/process.dart';
+import '../base/template.dart';
+import '../base/user_messages.dart';
+import '../cache.dart';
+import '../convert.dart';
+import '../macos/xcode.dart';
+import '../template.dart';
+
+/// A class to handle interacting with Xcode via OSA (Open Scripting Architecture)
+/// Scripting to debug Flutter applications.
+class XcodeDebug {
+  XcodeDebug({
+    required Logger logger,
+    required ProcessManager processManager,
+    required Xcode xcode,
+    required FileSystem fileSystem,
+    required UserMessages userMessages,
+    String? flutterRoot,
+  })  : _logger = logger,
+        _processUtils = ProcessUtils(logger: logger, processManager: processManager),
+        _xcode = xcode,
+        _fileSystem = fileSystem,
+        _userMessage = userMessages,
+        _flutterRoot = flutterRoot ?? Cache.flutterRoot!;
+
+  final ProcessUtils _processUtils;
+  final Logger _logger;
+  final Xcode _xcode;
+  final FileSystem _fileSystem;
+  final UserMessages _userMessage;
+  final String _flutterRoot;
+
+  /// Process to start Xcode's debug action.
+  @visibleForTesting
+  Process? startDebugActionProcess;
+
+  /// Information about the project that is currently being debugged.
+  @visibleForTesting
+  XcodeDebugProject? currentDebuggingProject;
+
+  /// Whether the debug action has been started.
+  bool get debugStarted => currentDebuggingProject != null;
+
+  String get pathToXcodeApp {
+    // If the Xcode Select Path is /Applications/Xcode.app/Contents/Developer,
+    // the path to Xcode App is /Applications/Xcode.app
+
+    final String? pathToXcode = _xcode.xcodeSelectPath;
+    if (pathToXcode == null || pathToXcode.isEmpty) {
+      throwToolExit(_userMessage.xcodeMissing);
+    }
+    final int index = pathToXcode.indexOf('.app');
+    if (index == -1) {
+      throwToolExit(_userMessage.xcodeMissing);
+    }
+    return pathToXcode.substring(0, index + 4);
+  }
+
+  String get pathToXcodeAutomationScript {
+    final String flutterToolsAbsolutePath = _fileSystem.path.join(
+      _flutterRoot,
+      'packages',
+      'flutter_tools',
+    );
+    return '$flutterToolsAbsolutePath/bin/xcode_debug.js';
```

Comment:
what do you think about validating here that the file exists? from reading this code, it seems to me like if we somehow get the path wrong, it wouldn't crash until we get a (possibly cryptic) crash from `osascript`.

Reply:
It's not really an "artifact" so that may not make sense to use `getArtifactPath`... 
couple ideas:
1) Perhaps as a stopgap, we pass the flutter root up through the [global XCDevice context](https://github.com/flutter/flutter/blob/9cc4f94397f1f4b4d49ff38a02e5a0708dc9d68e/packages/flutter_tools/lib/src/context_runner.dart#L363-L378). And add a context override for it internally.

2) Or perhaps we make a non-static getter for [flutterRoot in the Cache class](https://github.com/flutter/flutter/blob/9cc4f94397f1f4b4d49ff38a02e5a0708dc9d68e/packages/flutter_tools/lib/src/cache.dart#L205)
Maybe something like
```dart
  String? get cachedFlutterRoot {
    return flutterRoot;
  }
```

And that can be overriden internally [via context](https://github.com/flutter/flutter/blob/9cc4f94397f1f4b4d49ff38a02e5a0708dc9d68e/packages/flutter_tools/lib/src/context_runner.dart#L155-L161), which already has an internal override.

Reply:
Synced with @chingjun on chat, we decided to move `pathToXcodeAutomationScript` to the Xcode class so it can be overriden via context easily.

Also, want to note, I renamed it `pathToXcodeAutomationScript` -> `xcodeAutomationScriptPath`